File: workers/api/src/handlers/outcomes.ts
/**
 * Outcome Feedback API handlers for ZORA CORE
 * 
 * Outcome Feedback & Continual Learning v1.0 (Iteration 00D6)
 * 
 * These endpoints provide outcome feedback management for ZORA entities:
 * - Record feedback on missions, workflows, projects, etc.
 * - Fetch feedback for specific targets
 * - Compute basic stats per target and per type
 */

import { Hono } from 'hono';
import type { AuthAppEnv } from '../middleware/auth';
import { getSupabaseClient } from '../lib/supabase';
import { jsonResponse, errorResponse } from '../lib/response';
import type { AuthContext } from '../lib/auth';

const outcomesHandler = new Hono<AuthAppEnv>();

const VALID_TARGET_TYPES = [
  'climate_mission',
  'workflow_run',
  'zora_shop_project',
  'foundation_project',
  'goes_green_profile',
  'goes_green_action',
  'academy_learning_path',
  'academy_lesson',
];

const VALID_SOURCES = ['user', 'agent', 'system', 'admin'];

const VALID_SENTIMENTS = ['very_positive', 'positive', 'neutral', 'negative', 'very_negative'];

interface FeedbackRecord {
  id: string;
  tenant_id: string;
  user_id: string | null;
  source: string;
  target_type: string;
  target_id: string;
  rating: number | null;
  sentiment: string | null;
  tags: string[] | null;
  comment: string | null;
  context: Record<string, unknown>;
  created_at: string;
}

interface FeedbackStats {
  target_type: string;
  target_id: string | null;
  count: number;
  avg_rating: number | null;
  sentiment_counts: Record<string, number>;
  top_tags: Array<{ tag: string; count: number }>;
  last_feedback_at: string | null;
  best_entities?: Array<{ target_id: string; avg_rating: number; feedback_count: number }>;
  worst_entities?: Array<{ target_id: string; avg_rating: number; feedback_count: number }>;
}

/**
 * POST /api/outcomes/feedback
 * Record feedback for a ZORA entity
 */
outcomesHandler.post('/feedback', async (c) => {
  const auth = c.get('auth') as AuthContext | undefined;
  
  if (!auth) {
    return errorResponse('UNAUTHORIZED', 'Authentication required', 401);
  }
  
  const body = await c.req.json().catch(() => ({}));
  const {
    target_type,
    target_id,
    rating,
    sentiment,
    tags,
    comment,
    context,
    source = 'user',
  } = body;
  
  if (!target_type || !target_id) {
    return errorResponse('VALIDATION_ERROR', 'target_type and target_id are required', 400);
  }
  
  if (!VALID_TARGET_TYPES.includes(target_type)) {
    return errorResponse('VALIDATION_ERROR', `Invalid target_type. Must be one of: ${VALID_TARGET_TYPES.join(', ')}`, 400);
  }
  
  if (!VALID_SOURCES.includes(source)) {
    return errorResponse('VALIDATION_ERROR', `Invalid source. Must be one of: ${VALID_SOURCES.join(', ')}`, 400);
  }
  
  if (rating !== undefined && rating !== null && (rating < 1 || rating > 5)) {
    return errorResponse('VALIDATION_ERROR', 'Rating must be between 1 and 5', 400);
  }
  
  if (sentiment && !VALID_SENTIMENTS.includes(sentiment)) {
    return errorResponse('VALIDATION_ERROR', `Invalid sentiment. Must be one of: ${VALID_SENTIMENTS.join(', ')}`, 400);
  }
  
  const supabase = getSupabaseClient(c.env);
  
  try {
    const feedbackData: Record<string, unknown> = {
      tenant_id: auth.tenantId,
      user_id: auth.userId,
      source,
      target_type,
      target_id,
      context: context || {},
    };
    
    if (rating !== undefined && rating !== null) {
      feedbackData.rating = rating;
    }
    if (sentiment) {
      feedbackData.sentiment = sentiment;
    }
    if (tags && Array.isArray(tags)) {
      feedbackData.tags = tags;
    }
    if (comment) {
      feedbackData.comment = comment;
    }
    
    const { data: feedback, error } = await supabase
      .from('outcome_feedback')
      .insert(feedbackData)
      .select()
      .single();
    
    if (error) {
      console.error('Error inserting feedback:', error);
      return errorResponse('DATABASE_ERROR', 'Failed to record feedback', 500);
    }
    
    // Create journal entry
    try {
      await supabase.from('journal_entries').insert({
        tenant_id: auth.tenantId,
        category: 'learning',
        event_type: 'outcome_feedback_recorded',
        title: `Feedback recorded for ${target_type}`,
        content: `New ${source} feedback recorded for ${target_type} (${target_id})`,
        metadata: {
          target_type,
          target_id,
          rating,
          source,
        },
      });
    } catch (journalError) {
      console.warn('Failed to create journal entry:', journalError);
    }
    
    return jsonResponse({ data: feedback, error: null });
  } catch (err) {
    console.error('Error recording feedback:', err);
    return errorResponse('INTERNAL_ERROR', 'Failed to record feedback', 500);
  }
});

/**
 * GET /api/outcomes/feedback
 * Get feedback for a specific target
 * 
 * Query params:
 * - target_type (required)
 * - target_id (required)
 * - limit (optional, default 50)
 * - offset (optional, default 0)
 */
outcomesHandler.get('/feedback', async (c) => {
  const auth = c.get('auth') as AuthContext | undefined;
  
  if (!auth) {
    return errorResponse('UNAUTHORIZED', 'Authentication required', 401);
  }
  
  const targetType = c.req.query('target_type');
  const targetId = c.req.query('target_id');
  const limit = parseInt(c.req.query('limit') || '50', 10);
  const offset = parseInt(c.req.query('offset') || '0', 10);
  
  if (!targetType || !targetId) {
    return errorResponse('VALIDATION_ERROR', 'target_type and target_id are required', 400);
  }
  
  if (!VALID_TARGET_TYPES.includes(targetType)) {
    return errorResponse('VALIDATION_ERROR', `Invalid target_type. Must be one of: ${VALID_TARGET_TYPES.join(', ')}`, 400);
  }
  
  const supabase = getSupabaseClient(c.env);
  
  try {
    const { data: feedback, error, count } = await supabase
      .from('outcome_feedback')
      .select('*', { count: 'exact' })
      .eq('tenant_id', auth.tenantId)
      .eq('target_type', targetType)
      .eq('target_id', targetId)
      .order('created_at', { ascending: false })
      .range(offset, offset + limit - 1);
    
    if (error) {
      console.error('Error fetching feedback:', error);
      return errorResponse('DATABASE_ERROR', 'Failed to fetch feedback', 500);
    }
    
    return jsonResponse({
      data: feedback || [],
      pagination: {
        total: count || 0,
        limit,
        offset,
        has_more: offset + (feedback?.length || 0) < (count || 0),
      },
      error: null,
    });
  } catch (err) {
    console.error('Error fetching feedback:', err);
    return errorResponse('INTERNAL_ERROR', 'Failed to fetch feedback', 500);
  }
});

/**
 * GET /api/outcomes/stats/target
 * Get stats for a specific target
 * 
 * Query params:
 * - target_type (required)
 * - target_id (required)
 */
outcomesHandler.get('/stats/target', async (c) => {
  const auth = c.get('auth') as AuthContext | undefined;
  
  if (!auth) {
    return errorResponse('UNAUTHORIZED', 'Authentication required', 401);
  }
  
  const targetType = c.req.query('target_type');
  const targetId = c.req.query('target_id');
  
  if (!targetType || !targetId) {
    return errorResponse('VALIDATION_ERROR', 'target_type and target_id are required', 400);
  }
  
  if (!VALID_TARGET_TYPES.includes(targetType)) {
    return errorResponse('VALIDATION_ERROR', `Invalid target_type. Must be one of: ${VALID_TARGET_TYPES.join(', ')}`, 400);
  }
  
  const supabase = getSupabaseClient(c.env);
  
  try {
    const { data: feedback, error } = await supabase
      .from('outcome_feedback')
      .select('*')
      .eq('tenant_id', auth.tenantId)
      .eq('target_type', targetType)
      .eq('target_id', targetId);
    
    if (error) {
      console.error('Error fetching feedback:', error);
      return errorResponse('DATABASE_ERROR', 'Failed to compute stats', 500);
    }
    
    const rows = feedback || [];
    
    if (rows.length === 0) {
      return jsonResponse({
        data: {
          target_type: targetType,
          target_id: targetId,
          count: 0,
          avg_rating: null,
          sentiment_counts: {},
          top_tags: [],
          last_feedback_at: null,
        } as FeedbackStats,
        error: null,
      });
    }
    
    // Compute stats
    const ratings = rows.filter(r => r.rating !== null).map(r => r.rating as number);
    const avgRating = ratings.length > 0 ? ratings.reduce((a, b) => a + b, 0) / ratings.length : null;
    
    const sentimentCounts: Record<string, number> = {};
    for (const row of rows) {
      if (row.sentiment) {
        sentimentCounts[row.sentiment] = (sentimentCounts[row.sentiment] || 0) + 1;
      }
    }
    
    const tagCounts: Record<string, number> = {};
    for (const row of rows) {
      if (row.tags && Array.isArray(row.tags)) {
        for (const tag of row.tags) {
          tagCounts[tag] = (tagCounts[tag] || 0) + 1;
        }
      }
    }
    
    const topTags = Object.entries(tagCounts)
      .sort((a, b) => b[1] - a[1])
      .slice(0, 10)
      .map(([tag, count]) => ({ tag, count }));
    
    const lastFeedbackAt = rows.reduce((max, row) => {
      return row.created_at > max ? row.created_at : max;
    }, rows[0].created_at);
    
    const stats: FeedbackStats = {
      target_type: targetType,
      target_id: targetId,
      count: rows.length,
      avg_rating: avgRating !== null ? Math.round(avgRating * 100) / 100 : null,
      sentiment_counts: sentimentCounts,
      top_tags: topTags,
      last_feedback_at: lastFeedbackAt,
    };
    
    return jsonResponse({ data: stats, error: null });
  } catch (err) {
    console.error('Error computing stats:', err);
    return errorResponse('INTERNAL_ERROR', 'Failed to compute stats', 500);
  }
});

/**
 * GET /api/outcomes/stats/type
 * Get aggregated stats for all entities of a type
 * 
 * Query params:
 * - target_type (required)
 */
outcomesHandler.get('/stats/type', async (c) => {
  const auth = c.get('auth') as AuthContext | undefined;
  
  if (!auth) {
    return errorResponse('UNAUTHORIZED', 'Authentication required', 401);
  }
  
  const targetType = c.req.query('target_type');
  
  if (!targetType) {
    return errorResponse('VALIDATION_ERROR', 'target_type is required', 400);
  }
  
  if (!VALID_TARGET_TYPES.includes(targetType)) {
    return errorResponse('VALIDATION_ERROR', `Invalid target_type. Must be one of: ${VALID_TARGET_TYPES.join(', ')}`, 400);
  }
  
  const supabase = getSupabaseClient(c.env);
  
  try {
    const { data: feedback, error } = await supabase
      .from('outcome_feedback')
      .select('*')
      .eq('tenant_id', auth.tenantId)
      .eq('target_type', targetType);
    
    if (error) {
      console.error('Error fetching feedback:', error);
      return errorResponse('DATABASE_ERROR', 'Failed to compute stats', 500);
    }
    
    const rows = feedback || [];
    
    if (rows.length === 0) {
      return jsonResponse({
        data: {
          target_type: targetType,
          target_id: null,
          count: 0,
          avg_rating: null,
          sentiment_counts: {},
          top_tags: [],
          last_feedback_at: null,
          best_entities: [],
          worst_entities: [],
        } as FeedbackStats,
        error: null,
      });
    }
    
    // Compute overall stats
    const ratings = rows.filter(r => r.rating !== null).map(r => r.rating as number);
    const avgRating = ratings.length > 0 ? ratings.reduce((a, b) => a + b, 0) / ratings.length : null;
    
    const sentimentCounts: Record<string, number> = {};
    for (const row of rows) {
      if (row.sentiment) {
        sentimentCounts[row.sentiment] = (sentimentCounts[row.sentiment] || 0) + 1;
      }
    }
    
    const tagCounts: Record<string, number> = {};
    for (const row of rows) {
      if (row.tags && Array.isArray(row.tags)) {
        for (const tag of row.tags) {
          tagCounts[tag] = (tagCounts[tag] || 0) + 1;
        }
      }
    }
    
    const topTags = Object.entries(tagCounts)
      .sort((a, b) => b[1] - a[1])
      .slice(0, 10)
      .map(([tag, count]) => ({ tag, count }));
    
    const lastFeedbackAt = rows.reduce((max, row) => {
      return row.created_at > max ? row.created_at : max;
    }, rows[0].created_at);
    
    // Compute per-entity stats for best/worst
    const entityStats: Record<string, { ratings: number[]; count: number }> = {};
    for (const row of rows) {
      const tid = row.target_id;
      if (!entityStats[tid]) {
        entityStats[tid] = { ratings: [], count: 0 };
      }
      entityStats[tid].count += 1;
      if (row.rating !== null) {
        entityStats[tid].ratings.push(row.rating);
      }
    }
    
    const entityAvgs = Object.entries(entityStats)
      .filter(([_, stats]) => stats.ratings.length > 0)
      .map(([targetId, stats]) => ({
        target_id: targetId,
        avg_rating: Math.round((stats.ratings.reduce((a, b) => a + b, 0) / stats.ratings.length) * 100) / 100,
        feedback_count: stats.count,
      }));
    
    const bestEntities = [...entityAvgs]
      .sort((a, b) => b.avg_rating - a.avg_rating)
      .slice(0, 5);
    
    const worstEntities = [...entityAvgs]
      .sort((a, b) => a.avg_rating - b.avg_rating)
      .slice(0, 5);
    
    const stats: FeedbackStats = {
      target_type: targetType,
      target_id: null,
      count: rows.length,
      avg_rating: avgRating !== null ? Math.round(avgRating * 100) / 100 : null,
      sentiment_counts: sentimentCounts,
      top_tags: topTags,
      last_feedback_at: lastFeedbackAt,
      best_entities: bestEntities,
      worst_entities: worstEntities,
    };
    
    return jsonResponse({ data: stats, error: null });
  } catch (err) {
    console.error('Error computing stats:', err);
    return errorResponse('INTERNAL_ERROR', 'Failed to compute stats', 500);
  }
});

export default outcomesHandler;

File: zora_core/learning/outcomes.py
"""
ZORA CORE Outcome Feedback & Continual Learning v1.0

This module provides the outcome/critic service for ZORA CORE.
It supports:
- Recording feedback on ZORA entities (missions, workflows, projects, etc.)
- Fetching feedback for specific targets
- Computing basic stats and insights per target and per type
- Preparing data for future continual learning and optimization
"""

import os
import logging
from datetime import datetime
from typing import Optional, List, Dict, Any
from dataclasses import dataclass, field, asdict
from collections import Counter
from supabase import create_client, Client

logger = logging.getLogger(__name__)

VALID_TARGET_TYPES = [
    "climate_mission",
    "workflow_run",
    "zora_shop_project",
    "foundation_project",
    "goes_green_profile",
    "goes_green_action",
    "academy_learning_path",
    "academy_lesson",
]

VALID_SOURCES = ["user", "agent", "system", "admin"]

VALID_SENTIMENTS = ["very_positive", "positive", "neutral", "negative", "very_negative"]


@dataclass
class FeedbackRecord:
    id: str
    tenant_id: str
    user_id: Optional[str]
    source: str
    target_type: str
    target_id: str
    rating: Optional[int]
    sentiment: Optional[str]
    tags: Optional[List[str]]
    comment: Optional[str]
    context: Dict[str, Any]
    created_at: str


@dataclass
class FeedbackStats:
    target_type: str
    target_id: Optional[str]
    count: int
    avg_rating: Optional[float]
    sentiment_counts: Dict[str, int]
    top_tags: List[Dict[str, Any]]
    last_feedback_at: Optional[str]
    best_entities: Optional[List[Dict[str, Any]]] = None
    worst_entities: Optional[List[Dict[str, Any]]] = None


class OutcomeService:
    """Service for managing outcome feedback and computing insights."""

    def __init__(self, supabase_url: Optional[str] = None, supabase_key: Optional[str] = None):
        self.supabase_url = supabase_url or os.environ.get("SUPABASE_URL")
        self.supabase_key = supabase_key or os.environ.get("SUPABASE_SERVICE_ROLE_KEY")
        self._client: Optional[Client] = None

    @property
    def client(self) -> Client:
        if self._client is None:
            if not self.supabase_url or not self.supabase_key:
                raise ValueError("SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY must be set")
            self._client = create_client(self.supabase_url, self.supabase_key)
        return self._client

    def record_feedback(
        self,
        tenant_id: str,
        target_type: str,
        target_id: str,
        user_id: Optional[str] = None,
        source: str = "user",
        rating: Optional[int] = None,
        sentiment: Optional[str] = None,
        tags: Optional[List[str]] = None,
        comment: Optional[str] = None,
        context: Optional[Dict[str, Any]] = None,
    ) -> FeedbackRecord:
        """
        Record feedback for a ZORA entity.
        
        Args:
            tenant_id: The tenant ID
            target_type: Type of entity (e.g., 'climate_mission', 'workflow_run')
            target_id: ID of the target entity
            user_id: Optional user ID (None for agent/system feedback)
            source: Feedback source ('user', 'agent', 'system', 'admin')
            rating: Optional rating 1-5
            sentiment: Optional sentiment value
            tags: Optional list of tags
            comment: Optional comment text
            context: Optional context dict
            
        Returns:
            FeedbackRecord with the created feedback
            
        Raises:
            ValueError: If validation fails
        """
        if target_type not in VALID_TARGET_TYPES:
            raise ValueError(f"Invalid target_type: {target_type}. Must be one of {VALID_TARGET_TYPES}")
        
        if source not in VALID_SOURCES:
            raise ValueError(f"Invalid source: {source}. Must be one of {VALID_SOURCES}")
        
        if rating is not None and (rating < 1 or rating > 5):
            raise ValueError(f"Invalid rating: {rating}. Must be between 1 and 5")
        
        if sentiment is not None and sentiment not in VALID_SENTIMENTS:
            raise ValueError(f"Invalid sentiment: {sentiment}. Must be one of {VALID_SENTIMENTS}")

        data = {
            "tenant_id": tenant_id,
            "target_type": target_type,
            "target_id": target_id,
            "source": source,
            "context": context or {},
        }
        
        if user_id:
            data["user_id"] = user_id
        if rating is not None:
            data["rating"] = rating
        if sentiment:
            data["sentiment"] = sentiment
        if tags:
            data["tags"] = tags
        if comment:
            data["comment"] = comment

        result = self.client.table("outcome_feedback").insert(data).execute()
        
        if not result.data:
            raise RuntimeError("Failed to insert feedback")
        
        row = result.data[0]
        
        self._create_journal_entry(tenant_id, target_type, target_id, rating, source)
        
        return FeedbackRecord(
            id=row["id"],
            tenant_id=row["tenant_id"],
            user_id=row.get("user_id"),
            source=row["source"],
            target_type=row["target_type"],
            target_id=row["target_id"],
            rating=row.get("rating"),
            sentiment=row.get("sentiment"),
            tags=row.get("tags"),
            comment=row.get("comment"),
            context=row.get("context", {}),
            created_at=row["created_at"],
        )

    def _create_journal_entry(
        self,
        tenant_id: str,
        target_type: str,
        target_id: str,
        rating: Optional[int],
        source: str,
    ) -> None:
        """Create a journal entry for feedback recording."""
        try:
            self.client.table("journal_entries").insert({
                "tenant_id": tenant_id,
                "category": "learning",
                "event_type": "outcome_feedback_recorded",
                "title": f"Feedback recorded for {target_type}",
                "content": f"New {source} feedback recorded for {target_type} ({target_id})",
                "metadata": {
                    "target_type": target_type,
                    "target_id": target_id,
                    "rating": rating,
                    "source": source,
                },
            }).execute()
        except Exception as e:
            logger.warning(f"Failed to create journal entry: {e}")

    def get_feedback_for_target(
        self,
        tenant_id: str,
        target_type: str,
        target_id: str,
        limit: int = 50,
        offset: int = 0,
    ) -> Dict[str, Any]:
        """
        Get feedback entries for a specific target.
        
        Args:
            tenant_id: The tenant ID
            target_type: Type of entity
            target_id: ID of the target entity
            limit: Maximum number of results
            offset: Offset for pagination
            
        Returns:
            Dict with 'data' (list of feedback) and 'pagination' info
        """
        result = (
            self.client.table("outcome_feedback")
            .select("*")
            .eq("tenant_id", tenant_id)
            .eq("target_type", target_type)
            .eq("target_id", target_id)
            .order("created_at", desc=True)
            .range(offset, offset + limit - 1)
            .execute()
        )
        
        count_result = (
            self.client.table("outcome_feedback")
            .select("id", count="exact")
            .eq("tenant_id", tenant_id)
            .eq("target_type", target_type)
            .eq("target_id", target_id)
            .execute()
        )
        
        total = count_result.count if count_result.count else 0
        
        feedback_list = [
            FeedbackRecord(
                id=row["id"],
                tenant_id=row["tenant_id"],
                user_id=row.get("user_id"),
                source=row["source"],
                target_type=row["target_type"],
                target_id=row["target_id"],
                rating=row.get("rating"),
                sentiment=row.get("sentiment"),
                tags=row.get("tags"),
                comment=row.get("comment"),
                context=row.get("context", {}),
                created_at=row["created_at"],
            )
            for row in result.data
        ]
        
        return {
            "data": [asdict(f) for f in feedback_list],
            "pagination": {
                "total": total,
                "limit": limit,
                "offset": offset,
                "has_more": offset + len(feedback_list) < total,
            },
        }

    def compute_basic_stats_for_target(
        self,
        tenant_id: str,
        target_type: str,
        target_id: str,
    ) -> FeedbackStats:
        """
        Compute basic stats for a specific target.
        
        Args:
            tenant_id: The tenant ID
            target_type: Type of entity
            target_id: ID of the target entity
            
        Returns:
            FeedbackStats with aggregated metrics
        """
        result = (
            self.client.table("outcome_feedback")
            .select("*")
            .eq("tenant_id", tenant_id)
            .eq("target_type", target_type)
            .eq("target_id", target_id)
            .execute()
        )
        
        rows = result.data or []
        
        if not rows:
            return FeedbackStats(
                target_type=target_type,
                target_id=target_id,
                count=0,
                avg_rating=None,
                sentiment_counts={},
                top_tags=[],
                last_feedback_at=None,
            )
        
        ratings = [r["rating"] for r in rows if r.get("rating") is not None]
        avg_rating = sum(ratings) / len(ratings) if ratings else None
        
        sentiment_counts: Dict[str, int] = Counter()
        for row in rows:
            if row.get("sentiment"):
                sentiment_counts[row["sentiment"]] += 1
        
        tag_counts: Dict[str, int] = Counter()
        for row in rows:
            if row.get("tags"):
                for tag in row["tags"]:
                    tag_counts[tag] += 1
        
        top_tags = [
            {"tag": tag, "count": count}
            for tag, count in tag_counts.most_common(10)
        ]
        
        last_feedback_at = max(row["created_at"] for row in rows)
        
        return FeedbackStats(
            target_type=target_type,
            target_id=target_id,
            count=len(rows),
            avg_rating=round(avg_rating, 2) if avg_rating else None,
            sentiment_counts=dict(sentiment_counts),
            top_tags=top_tags,
            last_feedback_at=last_feedback_at,
        )

    def compute_basic_stats_for_type(
        self,
        tenant_id: str,
        target_type: str,
        include_entities: bool = True,
    ) -> FeedbackStats:
        """
        Compute aggregated stats for all entities of a given type.
        
        Args:
            tenant_id: The tenant ID
            target_type: Type of entity
            include_entities: Whether to include best/worst entities
            
        Returns:
            FeedbackStats with aggregated metrics across all entities
        """
        result = (
            self.client.table("outcome_feedback")
            .select("*")
            .eq("tenant_id", tenant_id)
            .eq("target_type", target_type)
            .execute()
        )
        
        rows = result.data or []
        
        if not rows:
            return FeedbackStats(
                target_type=target_type,
                target_id=None,
                count=0,
                avg_rating=None,
                sentiment_counts={},
                top_tags=[],
                last_feedback_at=None,
                best_entities=[],
                worst_entities=[],
            )
        
        ratings = [r["rating"] for r in rows if r.get("rating") is not None]
        avg_rating = sum(ratings) / len(ratings) if ratings else None
        
        sentiment_counts: Dict[str, int] = Counter()
        for row in rows:
            if row.get("sentiment"):
                sentiment_counts[row["sentiment"]] += 1
        
        tag_counts: Dict[str, int] = Counter()
        for row in rows:
            if row.get("tags"):
                for tag in row["tags"]:
                    tag_counts[tag] += 1
        
        top_tags = [
            {"tag": tag, "count": count}
            for tag, count in tag_counts.most_common(10)
        ]
        
        last_feedback_at = max(row["created_at"] for row in rows)
        
        best_entities: List[Dict[str, Any]] = []
        worst_entities: List[Dict[str, Any]] = []
        
        if include_entities:
            entity_stats: Dict[str, Dict[str, Any]] = {}
            for row in rows:
                tid = row["target_id"]
                if tid not in entity_stats:
                    entity_stats[tid] = {"ratings": [], "count": 0}
                entity_stats[tid]["count"] += 1
                if row.get("rating") is not None:
                    entity_stats[tid]["ratings"].append(row["rating"])
            
            entity_avgs = []
            for tid, stats in entity_stats.items():
                if stats["ratings"]:
                    avg = sum(stats["ratings"]) / len(stats["ratings"])
                    entity_avgs.append({
                        "target_id": tid,
                        "avg_rating": round(avg, 2),
                        "feedback_count": stats["count"],
                    })
            
            entity_avgs.sort(key=lambda x: x["avg_rating"], reverse=True)
            best_entities = entity_avgs[:5]
            
            entity_avgs.sort(key=lambda x: x["avg_rating"])
            worst_entities = entity_avgs[:5]
        
        return FeedbackStats(
            target_type=target_type,
            target_id=None,
            count=len(rows),
            avg_rating=round(avg_rating, 2) if avg_rating else None,
            sentiment_counts=dict(sentiment_counts),
            top_tags=top_tags,
            last_feedback_at=last_feedback_at,
            best_entities=best_entities,
            worst_entities=worst_entities,
        )

    def update_insight_for_target(
        self,
        tenant_id: str,
        target_type: str,
        target_id: str,
    ) -> Dict[str, Any]:
        """
        Update or create insight record for a specific target.
        
        Args:
            tenant_id: The tenant ID
            target_type: Type of entity
            target_id: ID of the target entity
            
        Returns:
            The updated insight record
        """
        stats = self.compute_basic_stats_for_target(tenant_id, target_type, target_id)
        
        stats_dict = {
            "count": stats.count,
            "avg_rating": stats.avg_rating,
            "sentiment_counts": stats.sentiment_counts,
            "tag_counts": {t["tag"]: t["count"] for t in stats.top_tags},
            "last_feedback_at": stats.last_feedback_at,
        }
        
        existing = (
            self.client.table("outcome_insights")
            .select("id")
            .eq("tenant_id", tenant_id)
            .eq("target_type", target_type)
            .eq("target_id", target_id)
            .eq("summary_type", "basic_stats")
            .execute()
        )
        
        if existing.data:
            result = (
                self.client.table("outcome_insights")
                .update({
                    "stats": stats_dict,
                    "updated_at": datetime.utcnow().isoformat(),
                })
                .eq("id", existing.data[0]["id"])
                .execute()
            )
        else:
            result = (
                self.client.table("outcome_insights")
                .insert({
                    "tenant_id": tenant_id,
                    "target_type": target_type,
                    "target_id": target_id,
                    "summary_type": "basic_stats",
                    "stats": stats_dict,
                })
                .execute()
            )
        
        return result.data[0] if result.data else {}

    def update_insight_for_type(
        self,
        tenant_id: str,
        target_type: str,
    ) -> Dict[str, Any]:
        """
        Update or create aggregated insight record for a target type.
        
        Args:
            tenant_id: The tenant ID
            target_type: Type of entity
            
        Returns:
            The updated insight record
        """
        stats = self.compute_basic_stats_for_type(tenant_id, target_type)
        
        stats_dict = {
            "count": stats.count,
            "avg_rating": stats.avg_rating,
            "sentiment_counts": stats.sentiment_counts,
            "tag_counts": {t["tag"]: t["count"] for t in stats.top_tags},
            "last_feedback_at": stats.last_feedback_at,
            "best_entities": stats.best_entities,
            "worst_entities": stats.worst_entities,
        }
        
        existing = (
            self.client.table("outcome_insights")
            .select("id")
            .eq("tenant_id", tenant_id)
            .eq("target_type", target_type)
            .is_("target_id", "null")
            .eq("summary_type", "basic_stats")
            .execute()
        )
        
        if existing.data:
            result = (
                self.client.table("outcome_insights")
                .update({
                    "stats": stats_dict,
                    "updated_at": datetime.utcnow().isoformat(),
                })
                .eq("id", existing.data[0]["id"])
                .execute()
            )
        else:
            result = (
                self.client.table("outcome_insights")
                .insert({
                    "tenant_id": tenant_id,
                    "target_type": target_type,
                    "target_id": None,
                    "summary_type": "basic_stats",
                    "stats": stats_dict,
                })
                .execute()
            )
        
        return result.data[0] if result.data else {}

    def refresh_all_insights_for_tenant(
        self,
        tenant_id: str,
    ) -> Dict[str, int]:
        """
        Refresh all insights for a tenant.
        
        Args:
            tenant_id: The tenant ID
            
        Returns:
            Dict with counts of updated insights
        """
        updated_targets = 0
        updated_types = 0
        
        for target_type in VALID_TARGET_TYPES:
            result = (
                self.client.table("outcome_feedback")
                .select("target_id")
                .eq("tenant_id", tenant_id)
                .eq("target_type", target_type)
                .execute()
            )
            
            target_ids = set(row["target_id"] for row in result.data) if result.data else set()
            
            for target_id in target_ids:
                self.update_insight_for_target(tenant_id, target_type, target_id)
                updated_targets += 1
            
            if target_ids:
                self.update_insight_for_type(tenant_id, target_type)
                updated_types += 1
        
        return {
            "updated_targets": updated_targets,
            "updated_types": updated_types,
        }
